<a href="https://colab.research.google.com/github/envgp/taking_the_pulse_of_the_planet/blob/main/module_0.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Introduction

Welcome to Module 0 of Taking the Pulse of the Planet class! Here you will find code snippets that will be useful in your assignments, and provides an in depth overview of the Python basics needed to get started with the class.

# Getting Started with Python
Python is a programming language, which is used to make your computer run algorithms! For our class, we'll be using Python to write code that will make our computer do work for us, such as process data and create charts. Let's get started with some of the basics of Python.
On each of the code cells below, hover over them and click on the triangle in the circle button to "run," or execute the cell. This tells our computer to go through and perform the actions we've asked it to do, through Python!

In [ ]:
print("Hello World")

Hello World


From that code, you should have seen an output right under it saying "Hello World" without the quotation marks. That's a print statement, and your first line of Python code.

What a print statement does is it asks the computer to display a line of text -- in this case, our line of text is Hello World. Notice how we ran our code with the Hello World in quotation marks. This is to inform the computer that we're trying to print a data structure known as a "string," or a line of text. In more general terms, it tells our computer that this statement within the quotes isn't code, but instead the message it should display!

print() itself is something known as a function. Similar to the functions you may have learned about in math, functions just take in an value ("hello world"), and perform some action, such as printing the input out in this case! We'll see lots of examples of functions later on, but the uniting feature of all of them is how they take input and perform some action. Some can even give you new values to work with!

Let's try assigning variables! Variables can be seen as boxes for values, where the values can be nearly anything! For example, we can assign numbers, strings, or whole objects to variables, which will then serve as a name for it.

One way to think about this is to think of your own name as a variable -- it's a way to refer to yourself, and that name could be reassigned to someone else as well. A rose by any other name would smell as sweet, right?

In [ ]:
x = 10
print(x)

10


From the code, your output should have been just the number 10. What you did was assign x, your variable name, the value 10, and then asked the computer to print out the value of x using print.

The number value 10 is also a data structure -- an int! Int stands for integer, and although we won't look too deeply into the differentiation between different number types, just know that in Python, you can perform numerical operations just as you generally would on a calculator.

In [ ]:
y = "I like cats!"
print(x)

10


Hey! This didn't print "I like cats!"

This is because we printed out the value of variable x instead of variable y. (Notice how the value of variables are still valid from cell to cell, once you've defined them!)

Try changing the input of the print statement to print our the value of our variable y. You can re-run the code cell as many times as you'd like before moving on.

Next, let's think about taking notes. Observe and run the code cell below:

In [ ]:
# This is a comment! I don't do anything

Huh! Nothing should have printed from that -- this is because that line of code was a comment. Comments tell the computer to skip over the line and ignore it.


Comments are great ways to make notes in your code, such as reminding yourself what each line of code does. While this class won't require you to comment your code, it can be a great way to stay on track. For this toolbox, and for many of the homework assignments for this class, these comments will serve as pointers and helpful hints for you to help guide you along.

In [ ]:
my_array = [1, 2, 3, 4, 5]
print(my_array)

[1, 2, 3, 4, 5]


And finally, let's look at arrays. Notice how I have multiple values in there separated by a comma. It's easy to think of arrays as lists of stuff, which can be numbers like our example, strings, other variables, or even other arrays! If our variables were boxes with something in it, arrays are simply just a variable pointing to a sequence of boxes, each of which has some value in it.

(For example, you could have an array called foods, where foods represents a sequence of objects, such as apples, ice cream, and sushi.)

# Packages
Packages include prewritten code someone else has prepared in a way that's easily portable -- just like you'd box up a present for a friend! This makes it easy to share code, and we'll be using quite a few packages so that we don't need to write a ton of things from scratch.

Different packages do different things, but for the purposes of our class, we'll be using a few important ones to help us with our work. Let's find out what they do!

## Pandas

No, not like the animal. (I wish...)

Pandas is a package used for data science, and it's use to manipulate data, especially in a table format. Like before, let's import pandas, but under the name pd.

In [ ]:
import pandas as pd

### Reading in Data

We have to get data from somewhere before we can analyze it. This command down here saves the data to our file system, so we can access it later. Don't worry too much about what git clone means at the moment, but notice how the link is where the data we want "lives" on the web.

For the purposes of this class, most of this data setup work will be done for you, and feel free to reach out if you're having any issues with it.

In [ ]:
!git clone https://premonition.stanford.edu/taking-the-pulse-of-the-planet/toolbox-data

Cloning into 'toolbox-data'...
remote: Enumerating objects: 3, done.
remote: Counting objects: 100% (3/3), done.
remote: Compressing objects: 100% (2/2), done.
remote: Total 3 (delta 0), reused 0 (delta 0), pack-reused 0
Receiving objects: 100% (3/3), done.


In order to tell our pandas module what data we're working with, we'll need something called a "path" to the location of the data from the file system we're working on here. For now, we'll be looking at a database about penguins!

This penguin data is stored in a "csv" file, which stands for comma-separated values. You don't need to worry about what these files do, but just think of it as a quick and handy way to store data in a tabular format within our files.

In [ ]:
# dataset_variable = pd.read_csv("path_to_csv_file")
penguins = pd.read_csv("./toolbox-data/penguins.csv")

Awesome! So now we have the variable "penguins" which contains all the data from the penguins.csv file we read in earlier.

Now that we've loaded in the CSV file, let's take a look at what it contains. We can use the info() function for this.

In [ ]:
# dataset_variable.info()
penguins.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 344 entries, 0 to 343
Data columns (total 9 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   rowid              344 non-null    int64  
 1   species            344 non-null    object 
 2   island             344 non-null    object 
 3   bill_length_mm     342 non-null    float64
 4   bill_depth_mm      342 non-null    float64
 5   flipper_length_mm  342 non-null    float64
 6   body_mass_g        342 non-null    float64
 7   sex                333 non-null    object 
 8   year               344 non-null    int64  
dtypes: float64(4), int64(2), object(3)
memory usage: 24.3+ KB


The info() function displays information about whatever dataset it's called on. Notice how Google Colab automatically displays all the information for you, even though you didn't print it all out. As it turns out, you don't always need an explicit print statement for things in Google Colab, as some functions will display things for you. We won't go too indepth on this however, because we got more interesting things to look at!

What did the info() function display? If you're looking at the output and seeing 344 entries, with a bunch of columns, aka "indexes" with different labels, you're on the right track.

But what about looking at the actual data of this dataset? The head() command lets us take a look at a few examples from the literal head, or front, of the table. Usually, this displays up to the first five results of your dataframe.

In [ ]:
# dataset_variable.head()
penguins.head()

rowid species     island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
0      1  Adelie  Torgersen            39.1           18.7              181.0   
1      2  Adelie  Torgersen            39.5           17.4              186.0   
2      3  Adelie  Torgersen            40.3           18.0              195.0   
3      4  Adelie  Torgersen             NaN            NaN                NaN   
4      5  Adelie  Torgersen            36.7           19.3              193.0   

   body_mass_g     sex  year  
0       3750.0    male  2007  
1       3800.0  female  2007  
2       3250.0  female  2007  
3          NaN     NaN  2007  
4       3450.0  female  2007

Look at all those chickens! I mean, penguins! We can see that there's a lot of different columns of data for this dataset.

###Operations
But what happens if we want to do stuff with the data? For instance, what if we wanted to look at only the data of the body mass of our penguins? What we can do is look at the indexes of our dataset -- focusing in on a particular column of data.

In [ ]:
# dataset_variable[index_name]
penguins["body_mass_g"]

0      3750.0
1      3800.0
2      3250.0
3         NaN
4      3450.0
        ...  
339    4000.0
340    3400.0
341    3775.0
342    4100.0
343    3775.0
Name: body_mass_g, Length: 344, dtype: float64

Wow! These are some chunky penguins, and we now have a big list of penguin body masses. But what if we wanted to find out the average/mean weight of our penguins? As it turns out, we can run operations on these columns. We won't go through all the operations right now, but know that they exist, and generally the code below is how we can run them on a column of data.

In [ ]:
# dataset_variable[index].operation()
penguins["body_mass_g"].mean()

4201.754385964912

What does the above code tell us about the average body mass of a penguin? Was this above or below your expectations?

But what if we're curious about the distribution of data given different labels. For example, let's say we want to group our penguins by their species, and find out the average body mass of the different species of penguins. Turns out, there's a function in pandas that makes this process easy -- it's called the groupby function.

This function allows us to generate new dataframes from the dataset we were already working on.

In [ ]:
# new_dataframe_var = dataset_variable.groupby(group_index)[info_index]
penguin_species = penguins.groupby("species")["body_mass_g"]

Alright, so now we have a new dataframe of penguin body masses, grouped by their species. This means that for all the body mass datapoints in our original dataset, they'll be sorted by the kind of species they are, into unique groups.

But let's try finding the average body mass, or mean of this set. What operation did we use before for finding the mean?

In [ ]:
# new_dataframe_var.operation()
penguin_species.mean()

species
Adelie       3700.662252
Chinstrap    3733.088235
Gentoo       5076.016260
Name: body_mass_g, dtype: float64

That should have given you a list of penguin species and their weights. By using these table/dataframe operations, we can manipulate our data into ways to calculate things important to us.

## Optional: Numpy

Numpy is one of the most common packages you'll encounter in data science -- so much so that many Python modules are actually dependent on Numpy to do tasks. For right now, our class will mainly focus on Pandas, so this section is currently optional.

First, let's import numpy.

In [ ]:
import numpy as np

This tells us that we should bring in the package named numpy, but under the name np. Think back to our variable naming skills, because the "as np" has us assign the package to the name np. In the future, whenever we want to reference or do something with numpy in this code, we'll be able to call upon it with np.

Remember that numpy is a collection of code -- you can find their documentation online [here](https://numpy.org/doc/stable/index.html)! For now, we'll just go over the basics.

Numpy is powerful because it helps us with processing data -- through the usage of numpy arrays. Don't worry too much about the differences between numpy arrays and python arrays, but feel free to explore them on your own time.

Numpy allows us to perform mathematical operations on numpy arrays. This means that we'll be able to use our numpy arrays to call the numpy functions we've imported.

In [ ]:
numbers = np.array([1, 2, 3, 4, 5, 6])
sum_value = numbers.sum()
print(sum_value)

21


The sum() function, is one imported by numpy. Like its name suggests, it adds up all the values of the numbers array. Our code saves it in the variable sum_value, which we then print.

If this is confusing, don't worry! We largely won't be working too much with numpy -- it's just good to know that it exists, because many data science libraries are built on top of it.